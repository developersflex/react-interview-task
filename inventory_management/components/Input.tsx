import { useField } from "formik";
import React from "react";

type InputProps = {
  placeholder: string;
  label: string;
  name: string;
  className?: string;
};

const Input: React.FC<InputProps> = ({
  placeholder,
  label,
  name,
  className,
}) => {
  const [field, meta] = useField(name);
  const hasError = meta.touched && meta.error;

  return (
    <div className={`w-full mb-5 ${className}`}>
      <div className="relative flex flex-col gap-1">
        <label className="ml-4 text-base font-semibold">{label}</label>
        <input
          type="text"
          id={label}
          placeholder={placeholder}
          className="rounded-md bg-brand-background-primary placeholder-[#E0E0E1] h-8 px-2 py-1"
          {...field}
        />
        {hasError && (
          <p className="text-brand-red-1000 text-xs ml-2 absolute -bottom-5">
            {meta.error}
          </p>
        )}
      </div>
    </div>
  );
};

export default Input;
