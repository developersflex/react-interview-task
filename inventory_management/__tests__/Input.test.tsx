import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { Form, Formik } from "formik";
import Input from "@/components/Input";
import { act } from "react-dom/test-utils";
import * as Yup from "yup";

describe("Input Component", () => {
  it("renders the input with the correct placeholder and label", () => {
    const placeholder = "Placeholder text";
    const label = "Label Text";
    const name = "Input Name";

    const { container, getByLabelText, getByPlaceholderText } = render(
      <Formik initialValues={{}} onSubmit={() => {}}>
        <Input placeholder={placeholder} label={label} name={name} />
      </Formik>
    );

    const inputElement = getByPlaceholderText(placeholder);
    const labelElement = getByLabelText(label);

    expect(inputElement).toBeInTheDocument();
    expect(labelElement).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it("allows user interaction", async () => {
    await act(async () => {
      render(
        <Formik initialValues={{ name: "" }} onSubmit={() => {}}>
          <Input label="Name" name="name" placeholder="Enter your name" />
        </Formik>
      );
    });

    const input = screen.getByPlaceholderText(
      "Enter your name"
    ) as HTMLInputElement;

    await act(async () => {
      fireEvent.change(input, { target: { value: "Durim" } });
    });

    expect(input.value).toBe("Durim");
  });

  it("displays error message when there are validation errors", async () => {
    const validationSchema = Yup.object().shape({
      testName: Yup.array().required("This field is required"),
    });

    const { getByText } = render(
      <Formik
        initialValues={{ testName: "" }}
        onSubmit={() => {}}
        validationSchema={validationSchema}
      >
        <Form>
          <Input label="Name" name="testName" placeholder="Enter your name" />
          <button type="submit">Submit</button>
        </Form>
      </Formik>
    );

    fireEvent.click(getByText("Submit"));

    await waitFor(async () => {
      const expectedErrorMessage = "This field is required";
      expect(getByText(expectedErrorMessage)).toBeInTheDocument();
    });
  });
});
