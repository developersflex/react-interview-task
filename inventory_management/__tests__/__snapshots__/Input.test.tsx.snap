// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component renders the input with the correct placeholder and label 1`] = `
<div>
  <div
    class="w-full mb-5 undefined"
  >
    <div
      class="relative flex flex-col gap-1"
    >
      <label
        class="ml-4 text-base font-semibold"
        for="Input Name"
      >
        Label Text
      </label>
      <input
        class="rounded-md bg-brand-background-primary placeholder-[#E0E0E1] h-8 px-2 py-1"
        id="Input Name"
        name="Input Name"
        placeholder="Placeholder text"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
